### PR TITLE
Energy evaluation cleanup ii

### PR DESCRIPTION
**Pull request by mtsch**
- `r_cut` is now computed from the boundary condition on the fly
- simplify energy evaluation
- fix bug with updating tangent matrices
- big test checking move consistency with various combinations of ensembles/potentials/boundary conditions (WIP)

### CODE REVIEW

File: src/energy_evaluation/abstract.jl [hunk 1]
```diff
@@ -61,11 +59,30 @@ abstract type AbstractPotentialVariables end
...
+- [`dimer_energy`](@ref)
 """
 abstract type AbstractDimerPotential <: AbstractPotential end
 
+"""
+    dimer_energy(potential::AbstractDimerPotential, r2::Real)
+    dimer_energy(potential::AbstractDimerPotentialB, r2::Real, z_angle::Real)
+
+Calculate the energy of dimer for given potential `potential` and squared distance `r2`
+between atoms. For an [`AbstractDimerPotentialB`](@ref), the angle between the line
```
Comment (ElkePahl) [inline src/energy_evaluation/abstract.jl]: ... the angle between the molecular axis and the z-direction ...

File: src/energy_evaluation/abstract.jl [hunk 2]
```diff
@@ -82,79 +99,72 @@ radius of the potential.
...
+            end
         end
     end
-    return dimer_energy_vec, energy_tot
+    return energy_tot + long_range_correction(config.boundary_condition, pot, num_atoms)
 end
-function dimer_energy_config(
-    distmat, NAtoms, potential_variables, r_cut, bc, pot::AbstractDimerPotential
+function dimer_energy_config!(
+    dimer_energy_vec, config, dist2_mat, pot::AbstractDimerPotential
```
Comment (ElkePahl) [inline src/energy_evaluation/abstract.jl]: perhaps rename dimer_energy_vec to en_atom_vec (fieldname in DimerPotentialVariables)

File: src/energy_evaluation/abstract.jl [hunk 3]
```diff
@@ -234,195 +235,41 @@ specific [`energy_update!`](@ref) function.
...
+        potential_variables.new_tan_vec .= (
+            get_tan(trial_pos, b, config.boundary_condition) for b in config
+        )
+        potential_variables.new_tan_vec[index] = 0
+
+        new_tan_vec = potential_variables.new_tan_vec
+        tan_mat = potential_variables.tan_mat
+
+        @views delta_en =
+            dimer_energy_atom(index, new_dist2_vec, new_tan_vec, cutoff, potential) -
```
Comment (ElkePahl) [inline src/energy_evaluation/abstract.jl]: As already discussed: for the TO DO list: make sure that stored energies per atom are used instead of recalculating the old energies/atom

File: src/energy_evaluation/LookupTablePotential.jl [hunk 4]
```diff
@@ -136,7 +135,7 @@ function set_variables(
     N = length(config)
     tan_matrix = get_tantheta_mat(config)
 
-    return LookupTableVariables{T}(zeros(N), tan_matrix, tan_matrix, zeros(N))
+    return LookupTableVariables{T}(zeros(N), tan_matrix, copy(tan_matrix), zeros(N))
```
Comment (ElkePahl) [inline src/energy_evaluation/LookupTablePotential.jl]: Was this the problem with tanmat?

File: src/BoundaryConditions.jl [hunk 5]
```diff
@@ -37,24 +51,40 @@ case of [`PeriodicBC`](@ref)), or return `nothing` if the position is invalid.
...
-    long_range_correction(bc::AbstractBC, potential, num_atoms, r_cut)
+    long_range_correction(bc::AbstractBC, potential, num_atoms)
     long_range_correction(potential, num_atoms, r_cut)
 
-Compute correction to energy from atoms outside the boundary condition. It is the integral
-of all interaction outside the cutoff distance, using uniform density approximation.
+Compute correction to energy from atoms outside the boundary condition, e.g. an integral
+of all interaction outside the cutoff distance using uniform density approximation.
 
 The first method should call the second an multiply it with an appropriate factor (for
```
Comment (ElkePahl) [inline src/BoundaryConditions.jl]: typo: an ->and

File: src/BoundaryConditions.jl [hunk 6]
```diff
@@ -37,24 +51,40 @@ case of [`PeriodicBC`](@ref)), or return `nothing` if the position is invalid.
...
 periodic boundary conditions) or return zero (for boundary conditions where a long range
 correction is not necessary).
 
-The second method only needs to be defined for a given potential if used with periodic
-boundary conditions.
+The second method only needs to be defined for a given potential for it to be usable with
+periodic boundary conditions.
```
Comment (ElkePahl) [inline src/BoundaryConditions.jl:65]: Do we still need the second method?

File: src/BoundaryConditions.jl [hunk 7]
```diff
@@ -119,21 +147,12 @@ Is abstract type for periodic boundary conditions to simulate bulk systems.
 
 In addition to the methods required by [`AbstractBC`](@ref), a `PeriodicBC` should
 implement
```
Comment (ElkePahl) [inline src/BoundaryConditions.jl]: It seems that something gone missing here?

File: src/energy_evaluation/abstract.jl [hunk 8]
```diff
@@ -61,100 +58,77 @@ abstract type AbstractPotentialVariables end
...
+# TODO: this function needs more work
 """
-    dimer_energy_config(distmat, num_atoms, potential_variables, pot)
-    dimer_energy_config(distmat, num_atoms, potential_variables, r_cut, pot)
-    dimer_energy_config(distmat, num_atoms, potential_variables, r_cut, boundary_condition, potential)
+    dimer_energy_config!(config, dist2_mat, potential_variables, potential)
 
-Stores the total of dimer energies of one atom with all other atoms in vector and calculates
-total energy of configuration.
+Store the dimer energies of one atom with all other atoms in `dimer_energies` and return
```
Comment (ElkePahl) [inline src/energy_evaluation/abstract.jl]: docu has to be updated including the function call

File: src/Ensembles.jl
Comment (ElkePahl) [inline src/Ensembles.jl]: Contains NPT ensemble variables that change during MC run.

File: src/Ensembles.jl [hunk 9]
```diff
@@ -93,36 +90,18 @@ NPT ensemble specific variable that change during MC run.
     -   `trial_move::SVector{3,T}`
     -   `trial_config::Config`
     -   `new_dist2_mat::Matrix{T}`
-    -   `r_cut::T`
-    -   `new_r_cut::T`
 When trialing a new configuration we select an atom at `index` to move to new position `trial_move`,
```
Comment (ElkePahl) [inline src/Ensembles.jl]: Using an $$NPT$$ ensemble, the type of move is selected according to `index`. For indices smaller or equal to the number of atoms in the system, a `trial_move` for the `index`-th atom is generated. If `index` is larger that the number of atoms, a volume move is trialled involving the generation of a scaled `trial_config` and corresponding `new_dist2_mat`.

File: src/Exchange.jl [hunk 10]
```diff
@@ -45,8 +45,8 @@ export parallel_tempering_exchange!, update_max_stepsize!
 # end
 """
     metropolis_condition(delta_energy::Number, beta::Number)
-    metropolis_condition(ensemble::Etype, delta_energy::Float64, volume_changed::Float64, volume_unchanged::Float64, beta::Float64) where Etype <: NPT
-    metropolis_condition(movetype::String, mc_state::MCState, ensemble::Etype) where Etype <: AbstractEnsemble
+    metropolis_condition(ensemble::NPT, delta_energy::Float64, volume_changed::Float64, volume_unchanged::Float64, beta::Float64)
+    metropolis_condition(movetype::String, mc_state::MCState, ensemble)
```
Comment (ElkePahl) [inline src/Exchange.jl:66]: I think that this is a bit chaotic. Would it make sense to separate last method from first two as they do different things. First two could be renamed to get_metropolis_probability or something like this.

Documentation has to be updated anyway as only three methods left.
Suggestions:
1) metropolis_condition(movetype ...) 
Returns probability for given `move_type` (atom, volume or atom swap moves). Perhaps provide formulae here?
2) get_metropolis_probability(...)
     get_metropolis_probability(...)
Return probability for atom or atom swap moves (1st method) or volume move(2nd method)

File: src/multihist_NPT.jl [hunk 11]
```diff
@@ -233,108 +231,27 @@ function multihistogram_NPT(
...
         cp[i] = (eenthalpy2 - eenthalpy^2) / (k * temp_result[i]^2)
+        vol[i] = evolume
     end
-    println("temperature array: ", temp_result)
-    println("heat capacity array: ", cp)
+    #println("temperature array: ", temp_result)
+    #println("heat capacity array: ", cp)
 
-    return temp_result, cp
+    return (; T=temp_result, C=cp, V=vol)
```
Comment (ElkePahl) [inline src/multihist_NPT.jl:254]: Does this mean that you can select the return values you want?

File: src/Sampling.jl [hunk 12]
```diff
@@ -20,8 +20,7 @@ using ..BoundaryConditions
 using ..EnergyEvaluation
 using ..Ensembles
 """
-    update_energy_tot(mc_states::MCStateVector, ensemble::Etype) where Etype <: AbstractEnsemble
-    update_energy_tot(mc_states::MCStateVector, ensemble::NPT)
+    update_energy_tot(mc_states::MCStateVector, ensemble)
 
 Function to update the current energy and energy squared values for coarse analysis of averages at the end. These are weighted according to the ensemble, and as such a method for each ensemble is required.
 Two methods avoids needless for-loops, where the JIT can save us computation time.
```
Comment (ElkePahl) [inline src/Sampling.jl]: There is only one method left

File: test/move-consistency.jl [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+# current configuration yields the same result as the updated values.
+using Test
+using ParallelTemperingMonteCarlo
+using StaticArrays
+using LinearAlgebra
+
+"""
+    mc_move_deterministic!(accept, mc_state, move_strat, pot, ensemble)
+
+Like `mc_move!`, but takes accepts the step is `accept ≡ true`. Returns move name for easier
```
Comment (ElkePahl) [inline test/move-consistency.jl]: ... but accepts the step if ..

**Review by ElkePahl — COMMENTED** [review]
Looks good. Only few suggestions for change - mainly documentation - and a few questions :)

**Review by ElkePahl — APPROVED** [review]
all fine
